### PR TITLE
Adding windows support

### LINKED ISSUE #43: Unsupported platform
I am using Win10. When I run `solc-select install`, it returns `Available versions to install:
Unsupported platform.`

### LINKED ISSUE #46: Validate inputted versions
Print out error message for user if they enter a solc version too large or too small. Both commands currently fail silently. 
```
~$ solc-select install 0.9.1
~$
solc-select install 0.2.1
~$
```

**Maximum** is in the `list.json` with `latestRelease` tag 
> "latestRelease": "0.8.2"

**Minimum** is different for each distribution: 
- For _Mac_, the minimum release is **0.3.6**.  https://github.com/ethereum/solc-bin/blob/gh-pages/macosx-amd64/list.json#L901
- For _Linux_, the true minimum release is **0.4.0**.  [`releases`](https://github.com/ethereum/solc-bin/blob/gh-pages/linux-amd64/list.json#L758) contains versions starting 0.4.10, but other binaries are available here: https://github.com/crytic/solc/tree/master/linux/amd64
- For _Windows_, the minimum release is **0.4.1**. https://github.com/ethereum/solc-bin/blob/gh-pages/windows-amd64/list.json#L849

### PR DESCRIPTION
**Pull request by 0xicingdeath**
This PR comprises of a series of breaking changes to solc-select, including: 
1. Validates user versions to ensure that they are not trying to install a version < minimum or > maximum. If so, solc-select will print an error (closes #46)
2. Implements storage refactor to use individual `solc-{version}` folders instead of just the `artifacts/` directory (https://github.com/crytic/solc-select/pull/58)
3. Adds support for Windows v0.4.5 - latest (closes #43)
4. Adds CI tests for different versions of Solidity across platforms

### CODE REVIEW

**Comment by NatalieChin80** [comment]
Thanks to @Xenomega for testing this 🎉 

- solc-select installs properly but errors out when trying to use `solc`
- Mac/Linux binaries are just binaries, but Windows is a zip file containing solc and 2 dlls: https://binaries.soliditylang.org/windows-amd64/list.json
- Need to add extra code specifically for Windows to unzip 
- Going to refactor the folder structure for Mac/Linux to use `/artifacts/solc<version>/binary` in the case of collisions (this PR is on pause for the refactoring)

**Comment by NatalieChin80** [comment]
With #58 done, moving on to ensure that this works on Windows. 

Note, according to the [list.json](https://binaries.soliditylang.org/windows-amd64/list.json): 
- For all solc versions `<= 0.7.1`: the hosted path is in a `zip` file. These need to be unzipped.
- For all solc versions `>0.7.1`: the file is an `.exe`

**Comment by NatalieChin80** [comment]
🥳

Yay, this works for the executables 🎉 
``` 
$ solc-select install 0.8.4
Installing '0.8.4'...
Version '0.8.4' installed.

$ solc foo.sol
Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
--> foo.sol
```

**Comment by NatalieChin80** [comment]
To support `<=0.7.1`, the following logic needs to be added to `install_artifacts()`: 
```
if platform is windows AND version is <= 0.7.1: 
   unzip the downloaded file in the `solc-select/artifacts/solc-{version}` directory
   rename `solc.exe` to `solc-{version}.exe` 
```

and that should round up the support for all solc versions of Windows! 🥳

**Comment by 0xicingdeath** [comment]
e25f406 now unzips and renames executable for older versions of windows properly 🎉 
```
vagrant@vagrant-10 MINGW64 ~/.solc-select/artifacts
$ solc-select use 0.4.25
Switched global version to 0.4.25

vagrant@vagrant-10 MINGW64 ~/.solc-select/artifacts
$ solc foo.sol
foo.sol:3:2: Warning: Defining constructors as functions with the same name as the contract is deprecated. Use "constructor(...) { ... }" instead.
        function Foo() {
 ^ (Relevant source part starts here and spans across multiple lines).
foo.sol:3:2: Warning: No visibility specified. Defaulting to "public".
        function Foo() {
 ^ (Relevant source part starts here and spans across multiple lines).
```

**Comment by 0xicingdeath** [comment]
According to the [json](https://binaries.soliditylang.org/windows-amd64/list.json),`0.4.3, 0.4.4` are missing for Windows, which is why they fail to install: 
```
    "0.4.5": "solc-windows-amd64-v0.4.5+commit.b318366e.zip",
    "0.4.2": "solc-windows-amd64-v0.4.2+commit.af6afb04.zip",
    "0.4.1": "solc-windows-amd64-v0.4.1+commit.4fc6fc2c.zip"
```
Moving this to #64 to be fixed in a separate PR.

**Comment by 0xicingdeath** [comment]
In addition to the lack of support, it also looks like some of the zip files are using nightly builds instead of the released version.

For a Solidity file with `pragma solidity 0.4.1`, it prints out the following error message: 

```
foo.sol:1:1: Error: Source file requires different compiler version (current compiler is 0.4.1-develop.2016.09.09+commit.4fc6fc2c.Windows.msvc - note that nightly builds are considered to be strictly less than the released version
pragma solidity 0.4.1;
^--------------------^
```
Same with `0.4.2`, and `0.4.15`.

---

Due to this issue and [the comment above](https://github.com/crytic/solc-select/pull/54#issuecomment-874173959), validating min/max versions will be merged into this commit.

**Comment by 0xicingdeath** [comment]
2166d48 add CI tests to ensure parity between Linux, Mac OS X, and Windows. See: 

```bash
solc-select use 0.4.5  &> /dev/null
solc ./scripts/solidity_tests/solc045_success.sol

if [[ $? != 0 ]]; then
  echo "FAILED: solc045_success" 
  exit 255
fi
echo "SUCCESS: solc045_success"

execute=$(solc ./scripts/solidity_tests/solc045_fail_compile.sol 2>&1)
if [[ "$execute" != *"Error: Expected token Semicolon got 'Function'"* ]]; then
  echo "FAILED: solc045_fail_compile"
  echo "$execute"
  exit 255
fi
echo "SUCCESS: solc045_fail_compile"

solc-select use 0.5.0  &> /dev/null
solc ./scripts/solidity_tests/solc050_success.sol

if [[ $? != 0 ]]; then
  echo "FAILED: solc050_success" $?
  exit 255
fi
echo "SUCCESS: solc050_success"

execute=$(solc ./scripts/solidity_tests/solc050_fail_compile.sol 2>&1)
if [[ "$execute" != *"Error: Functions are not allowed to have the same name as the contract."* ]]; then
  echo "FAILED: solc045_fail_compile"
  echo "$execute"
  exit 255
fi
echo "SUCCESS: solc045_fail_compile"
```
in scripts/travis_test_solc.sh

**Comment by 0xicingdeath** [comment]
40eb19a tests all breaking changes between Solidity versions 🎉  We now have parity between all 3 platforms. 🥳

**Comment by 0xicingdeath** [comment]
This PR is now ready for review. 🎉  3d789b4 pulls the `latestRelease` from binaries.soliditylang to automate testing of the latest release for all platforms. I'd also like to expand the test suite, but that's more apt for a future PR.

```bash
latest_release=$(curl https://binaries.soliditylang.org/windows-amd64/list.json |
  python3 -c "import sys,json; print(json.load(sys.stdin)['latestRelease'])")
use_version=$(solc-select use "$latest_release")
if [[ $use_version != "Switched global version to $latest_release" ]]; then
  echo "WINDOWS FAILED: maximum version"
  exit 255
fi
echo "WINDOWS SUCCESS: maximum version"

use_version=$(solc-select use 0.100.8 2>&1)
if [[ $use_version != *"Invalid version '$latest_release' is the latest available version"* ]]; then
  echo "WINDOWS FAILED: version too high"
  exit 255
fi
```

File: scripts/travis_test_solc.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  exit 255
+fi
+echo "SUCCESS: solc060_success_receive"
+
+## solc 0.7.0 ##
+solc-select use 0.7.0  &> /dev/null
+
+execute=$(solc ./scripts/solidity_tests/solc070_fail_compile.sol 2>&1)
+if [[ "$execute" != *"has been deprecated."* ]]; then
+  echo "FAILED: solc070_fail_compile" $execute
```
Comment (arcz) [inline scripts/travis_test_solc.sh]: Double quote `$execute`:
`echo "FAILED: solc070_fail_compile" "$execute"`

File: solc_select/solc_select.py [hunk 2]
```diff
@@ -90,8 +137,23 @@ def switch_global_version(version):
 def valid_version(version):
     match = re.search(r"^(\d+).(\d+).(\d+)$", version)
 
+    url = f"https://binaries.soliditylang.org/{soliditylang_platform()}/list.json"
+    list_json = urllib.request.urlopen(url).read()
+    latest_release = json.loads(list_json)["latestRelease"]
+
```
Comment (arcz) [inline solc_select/solc_select.py]: only the `if StrictVersion(version) > StrictVersion(latest_release)` branch depends on this, can be moved there

File: solc_select/solc_select.py [hunk 3]
```diff
@@ -66,12 +101,24 @@ def is_older_linux(version):
...
+    )
+
+
 def get_url(version, artifact):
     if is_older_linux(version):
         return f"https://raw.githubusercontent.com/crytic/solc/master/linux/amd64/{artifact}"
     return f"https://binaries.soliditylang.org/{soliditylang_platform()}/{artifact}"
 
 
+def get_artifact(version, artifact_file_dir):
```
Comment (arcz) [inline solc_select/solc_select.py]: this function is unused, is this intentional?

Comment (0xicingdeath) [reply solc_select/solc_select.py]: Good catch! Removed in d583428

File: solc_select/solc_select.py (hunk 2 above)
Comment (0xicingdeath) [reply solc_select/solc_select.py]: Changed to validate regex first, then minimum version, then maximum version in d583428

File: scripts/travis_test_solc.sh (hunk 1 above)
Comment (0xicingdeath) [reply scripts/travis_test_solc.sh]: Also fixed in d583428

**Pull request merged**